exporter/perf_map.go

```diff
@@ -0,0 +1,99 @@
+package exporter
+
+import (
+	"log"
+	"time"
+
+	"github.com/cloudflare/ebpf_exporter/config"
+	"github.com/cloudflare/ebpf_exporter/decoder"
+	"github.com/iovisor/gobpf/bcc"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type PerfMapSink struct {
+	counterConfig config.Counter
+	counterVec    *prometheus.CounterVec
+}
+
+func NewPerfMapSink(decoders *decoder.Set, module *bcc.Module, counterConfig config.Counter) *PerfMapSink {
+	var (
+		receiveCh = make(chan []byte)
+		lostCh    = make(chan uint64)
+	)
+
+	sink := &PerfMapSink{
+		counterConfig: counterConfig,
+	}
+	sink.resetCounterVec()
+
+	table := bcc.NewTable(module.TableId(counterConfig.PerfMap), module)
+
+	perfMap, err := bcc.InitPerfMap(table, receiveCh, lostCh)
+	if err != nil {
+		log.Panicf("Can't init PerfMap: %s", err)
+	}
+
+	go func(sink *PerfMapSink, counterConfig config.Counter, receiveCh <-chan []byte) {
+		for rawBytes := range receiveCh {
+			// https://github.com/cilium/ebpf/pull/94#discussion_r425823371
+			// https://lore.kernel.org/patchwork/patch/1244339/
+			var validDataSize uint
+			for _, labelConfig := range counterConfig.Labels {
+				validDataSize += labelConfig.Size
+			}
+
+			labelValues, err := decoders.DecodeLabels(rawBytes[:validDataSize], counterConfig.Labels)
+			if err != nil {
+				if err == decoder.ErrSkipLabelSet {
+					continue
+				}
+
+				log.Printf("failed to decode labels: %s", err)
+			}
+
+			sink.counterVec.WithLabelValues(labelValues...).Inc()
+
+		}
+	}(sink, counterConfig, receiveCh)
+
+	go func(sink *PerfMapSink) {
```

Comment:
Why do we need to reset counters every hour here? Should the time be configurable?

Reply:
For the same reason as using a perf map: to avoid unbound label cardinality. I could switch this from `time.Duration` to counting currently active counters and drop them once they cross a threshold. Should be easier to forecast growth.